File: 14. Delete node from BST.py
"""
Task: Delete node from BST
Time Complexity: O(h)
Space Complexity: O(h) — recursion stack
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def find_min_node(root):
    while root.left: root=root.left
    return root

def delete_node(root,key):
    if not root: return None
    if key<root.data:
        root.left=delete_node(root.left,key)
    elif key>root.data:
        root.right=delete_node(root.right,key)
    else:
        if not root.left:
            return root.right
        if not root.right:
            return root.left
        succ=find_min_node(root.right)
        root.data=succ.data
        root.right=delete_node(root.right,succ.data)
    return root

def inorder(root):
    if not root: return
    inorder(root.left); print(root.data, end=' '); inorder(root.right)

if __name__=='__main__':
    root=None
    root=Node(5); root.left=Node(3); root.right=Node(7)
    root=delete_node(root,3)
    print('After delete:', end=' '); inorder(root); print()

File: 14. Delete node from BST.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
Node* findMin(Node* root)
{
    while (root->left) root = root->left;
    return root;
}
Node* deleteNode(Node* root, int key)
{
    if (!root)
        return nullptr;
    if (key < root->data)
        root->left = deleteNode(root->left, key);
    else if (key > root->data)
        root->right = deleteNode(root->right, key);
    else
    {
        if (!root->left)
        {
            Node* r = root->right;
            delete root;
            return r;
        }
        else if (!root->right)
        {
            Node* l = root->left;
            delete root;
            return l;
        }
        else
        {
            Node* succ = findMin(root->right);
            root->data = succ->data;
            root->right = deleteNode(root->right, succ->data);
        }
    }
    return root;
}
void inorder(Node* root)
{
    if (!root)
        return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}
int main()
{
    Node* root = nullptr;
    root = new Node(5);
    root->left = new Node(3);
    root->right = new Node(7);
    root = deleteNode(root, 3);
    cout << "After delete: ";
    inorder(root);
    cout << endl;
    return 0;
}
